### PR TITLE
Add Nintendo 64 device mode support (N64Console)

### PR DESCRIPTION
**Pull request by JonnyHaystack**
Hi there, thanks for the great library. I have added Nintendo 64 console support and tested it in Goldeneye 007 on my N64. Everything seems to be working fine - it's a good bit simpler than the GCC protocol 😄

Shoutouts to houndsolo on discord for buying me an N64 for this, to Simple for answering a ton of questions, and to the reverse engineers behind those links you put in the wiki

### CODE REVIEW

File: src/GamecubeAPI.hpp [hunk 1]
```diff
@@ -24,6 +24,8 @@ THE SOFTWARE.
 // Include guard
 #pragma once
 
+#include "GamecubeAPI.h"
+
```
Comment (NicoHood) [inline src/GamecubeAPI.hpp:28]: Why did you add this include here?

File: src/N64.c [hunk 2]
```diff
@@ -60,4 +60,43 @@ bool n64_read(const uint8_t pin, N64_Report_t* report)
...
+    volatile uint8_t* outPort = portOutputRegister(port);
+    volatile uint8_t* inPort = portInputRegister(port);
+
+    // Don't want interrupts getting in the way
+    uint8_t oldSREG = SREG;
+    cli();
+
+    // Read in data from the console
+    // After receiving the init command you have max 20ms to respond (for the data command)!
+    uint8_t command[3] = {0, 0, 0};
```
Comment (NicoHood) [inline src/N64.c:82]: In the GC code I've added "TODO do not init" here. Do you know why I might have added this? Its been a looong time...

File: src/N64API.hpp [hunk 3]
```diff
@@ -24,6 +24,8 @@ THE SOFTWARE.
 // Include guard
 #pragma once
 
+#include "N64API.h"
```
Comment (NicoHood) [inline src/N64API.hpp:27]: Same here, why did you add that?

File: src/N64API.hpp [hunk 4]
```diff
@@ -154,4 +156,55 @@ N64_Data_t CN64Controller::getData(void)
...
+CN64Console::CN64Console(const uint8_t p) : pin(p){
+    // Empty
+}
+
+
+bool CN64Console::write(N64_Data_t &data)
+{
+    // Abort if controller was not initialized.
+    // N64 will refuse and weird connect/disconnect errors will occur.
+    //if (data.report.origin) {
```
Comment (NicoHood) [inline src/N64API.hpp:168]: Why did you disable this?

**Review by NicoHood — COMMENTED** [review]
Very nice PR, looks really clean!
Since I cannot test this and havent used it for quite some time, please answer my few questions, so I can merge that :-)

**Comment by NicoHood** [comment]
If you have any important information to add to the wiki, please let me know!

File: src/GamecubeAPI.hpp (hunk 1 above)
Comment (JonnyHaystack) [reply src/GamecubeAPI.hpp:28]: I was getting compilation errors without it when bundling with my project. I think it is generally best practice anyway to always include the header where its definitions are needed. Seeing as you have include guards/"pragma once" to prevent double inclusion, there's no reason not to, and this will ensure that definitions are always available by the time they are needed.

File: src/N64.c (hunk 2 above)
Comment (JonnyHaystack) [reply src/N64.c:82]: I was wondering the same myself I'm afraid :smile: Can't think what that would mean.. you definitely want to init the device haha

File: src/N64API.hpp (hunk 3 above)
Comment (JonnyHaystack) [reply src/N64API.hpp:27]: Same as above

File: src/N64API.hpp (hunk 4 above)
Comment (JonnyHaystack) [reply src/N64API.hpp:168]: N64 doesn't have a special origin report, it just has the initial status request command (0x00) and the report request/input poll (0x01). Sorry about leaving it commented out, I forgot to remove that. I'll delete it now

**Comment by JonnyHaystack** [comment]
> If you have any important information to add to the wiki, please let me know!

Thankfully all of the info needed was in the links you already put in the wiki, which were incredibly helpful. Only thing I'd suggest is making sure those aren't going to disappear. I'll make sure they're included in the internet archive.

**Comment by NicoHood** [comment]
I already inlined the information for n64, as I was also affraid that it disappeared.

File: src/GamecubeAPI.hpp (hunk 1 above)
Comment (NicoHood) [reply src/GamecubeAPI.hpp:28]: I think this is really not required, as it is included here:
https://github.com/NicoHood/Nintendo/blob/master/src/GamecubeAPI.h#L78

Can you please give it another test, I am very sure this is not required.

File: src/N64.c (hunk 2 above)
Comment (NicoHood) [reply src/N64.c:82]: I think my idea was to use `uint8_t command[3];`, since the initialization should not be required and takes time. However I am not sure, if an empty array is required. If not, we can remove the initialization. What do you think?

Comment (NicoHood) [reply src/N64.c:82]: Maybe you can try by setting it to `0xff, 0xff, 0xff` and check if the code still works.

File: src/N64API.hpp (hunk 4 above)
Comment (NicoHood) [reply src/N64API.hpp]: That makes sense. It should be removed then :-)

File: src/GamecubeAPI.hpp (hunk 1 above)
Comment (JonnyHaystack) [reply src/GamecubeAPI.hpp:28]: Hmm okay it does seem to be fine now. Maybe my includes in my project were wrong at the time. I can remove it if you want, although it does also give errors in my editor when the include isn't present.

Comment (NicoHood) [reply src/GamecubeAPI.hpp:28]: Yes, please remove it. It would be wrong to have the include there, since this is a header implementation, not a c file.

Comment (JonnyHaystack) [reply src/GamecubeAPI.hpp]: Okay, done

File: src/N64.c (hunk 2 above)
Comment (JonnyHaystack) [reply src/N64.c:82]: Will check this in a bit. Would've thought not initialising to 0 would result in invalid commands, because not all N64 commands are 3 bytes. In fact, the only ones that I use are 1-byte, so I imagine they could have junk data in them that causes the command to not be correctly recognised as 0x00 or 0x01. But maybe I'm wrong

Comment (JonnyHaystack) [reply src/N64.c:82]: Okay, it does seem to work without the initialisation (or with 3x 0xFF), so I'll keep that initialisation out.

Comment (NicoHood) [reply src/N64.c]: Do we even need 3 bytes then? What if we use 1 byte?

Can you please also correct this for gamecube?

Comment (JonnyHaystack) [reply src/N64.c]: Well, there are other commands which are 3 bytes. I just didn't implement those. They could be useful in future, although it's way more advanced

Comment (NicoHood) [reply src/N64.c]: Alright, thanks for the feedback! I think the PR is ready to merge then.

Comment (JonnyHaystack) [reply src/N64.c]: Ok that's done and tested now 😃

**Comment by NicoHood** [comment]
I've tagged a new release. Please give it a test and report any errors. Thanks for your help :-) Great work!
